src/main/java/com/gradle/enterprise/conventions/customvalueprovider/GradleEnterpriseConventions.java

```diff
@@ -127,17 +137,43 @@ public void setCommitId(File projectDir, BuildScanExtension buildScan, String co
 
         buildScan.value(GIT_COMMIT_NAME, commitId);
         customValueSearchUrl(Collections.singletonMap(GIT_COMMIT_NAME, commitId)).ifPresent(url -> buildScan.link("Git Commit Scans", url));
-        if (isCiServer) {
-            getRemoteGitHubRepository(projectDir).ifPresent(repoUrl -> buildScan.link("Source", String.format("%s/commit/%s", repoUrl, commitId)));
-        } else {
-            buildScan.background(__ -> getRemoteGitHubRepository(projectDir).ifPresent(repoUrl -> buildScan.link("Source", String.format("%s/commit/%s", repoUrl, commitId))));
-        }
+        // This is a configuration-safe way of invoking external process at the configuration time since Gradle 7.5.
+        // As of Gradle 8.7, it still may add the remote url to the configuration cache inputs when used in the
+        // buildScan.background callback, but this value doesn't change very often.
+        // CI value providers call this method outside the background callback, so the ValueSource is mandatory to avoid
+        // failing the build because of CC errors there.
+        @SuppressWarnings("UnstableApiUsage")
+        Provider<String> remoteGitHubRepositoryProvider = getProviderFactory().of(
```

Comment:
🤔 The `buildScan.background()` call should work, only you need to use the argument of the action instead of trying to add it to the captured buildScan extension. Check out what the CCUD plugin does: https://github.com/gradle/common-custom-user-data-gradle-plugin/blob/main/src/main/java/com/gradle/CustomBuildScanEnhancements.java#L379-L463

Reply:
> 🤔 The `buildScan.background()` call should work, only you need to use the argument of the action instead of trying to add it to the captured buildScan extension. Check out what the CCUD plugin does: https://github.com/gradle/common-custom-user-data-gradle-plugin/blob/main/src/main/java/com/gradle/CustomBuildScanEnhancements.java#L379-L463

Maybe the comment is not clear enough. There are two code paths that lead here:
1. local reporting that calls this method in the background callback
2. CI reporting that calls this method at configuration time

As far as I understood, (2) cannot use background callback for some reason (the proper link appears too late or something). This means that CI must wrap the external process in a value source to be CC-compatible, and the output becomes CC input, but this is intended - it is used at configuration time anyway.

(1) uses background callback. Direct call to `ProcessBuilder.start` would be fine there (and it works for CCUD), but for simplicity I keep using the value source too. ValueSources evaluated in background become configuration inputs because of the [existing issue](https://github.com/gradle/gradle/issues/25474 in Gradle). Given that the origin is unlikely to change often, it is probably fine, and the performance impact of validating the process' output is likely negligible.

Reply:
I am very surprised that (2) does not work as expected on CI when using background. Do you have an example for that?

I agree that having the repository as an configuration input will not cause many - if any - cache misses.